### ISSUE TITLE
Errors in Windows XP SP3

### ISSUE THREAD

**Issue by HunterZ** [rickparrish/FleckProxy#3]
I'm attempting to run the version dated Mar 17, 2015 in Windows XP SP3 inside of VirtualBox using the following command:

FleckProxy.exe -ws 1123

For some reason I am getting the following error:

4/26/2015 12:37:40 PM [Info] Starting server on port 1123
4/26/2015 12:37:40 PM [Info] Redirecting to localhost:23

4/26/2015 12:37:40 PM [Error] Unable to start server (SocketError=10047/AddressFamilyNotSupported: An address incompatible with the requested protocol was used)

Error bringing ws:// server online.  Server will now shutdown

If it helps, here is my ipconfig info:
Windows IP Configuration

        Host Name . . . . . . . . . . . . : marsbase
        Primary Dns Suffix  . . . . . . . : intertron
        Node Type . . . . . . . . . . . . : Unknown
        IP Routing Enabled. . . . . . . . : No
        WINS Proxy Enabled. . . . . . . . : No
        DNS Suffix Search List. . . . . . : intertron

Ethernet adapter Local Area Connection:

        Connection-specific DNS Suffix  . :
        Description . . . . . . . . . . . : AMD PCNET Family PCI Ethernet Adapte
r
        Physical Address. . . . . . . . . : 08-00-27-36-00-8B
        Dhcp Enabled. . . . . . . . . . . : Yes
        Autoconfiguration Enabled . . . . : Yes
        IP Address. . . . . . . . . . . . : 10.0.2.15
        Subnet Mask . . . . . . . . . . . : 255.255.255.0
        Default Gateway . . . . . . . . . : 10.0.2.2
        DHCP Server . . . . . . . . . . . : 10.0.2.2
        DNS Servers . . . . . . . . . . . : 192.168.1.1
        Lease Obtained. . . . . . . . . . : Sunday, April 26, 2015 7:49:30 AM
        Lease Expires . . . . . . . . . . : Monday, April 27, 2015 7:49:30 AM

On a hunch, I installed IPv6 support in case it was trying to bind IPv6 for the listen port. This gives me a new error:

4/26/2015 12:44:01 PM [Info] Starting server on port 1123
4/26/2015 12:44:01 PM [Info] Redirecting to localhost:23

4/26/2015 12:44:01 PM [Error] Unable to start server (SocketError=10042/Protocol
Option: An unknown, invalid, or unsupported option or level was specified in a g
etsockopt or setsockopt call) System.Net.Sockets.SocketException (0x80004005): A
n unknown, invalid, or unsupported option or level was specified in a getsockopt
 or setsockopt call
   at System.Net.Sockets.Socket.SetSocketOption(SocketOptionLevel optionLevel, S
ocketOptionName optionName, Int32 optionValue, Boolean silent)
   at System.Net.Sockets.Socket.SetSocketOption(SocketOptionLevel optionLevel, S
ocketOptionName optionName, Int32 optionValue)
   at Fleck.WebSocketServer..ctor(Int32 port, String location)
   at FleckProxy.Program.StartListener(Int32 port, String scheme)

Error bringing ws:// server online.  Server will now shutdown

**Comment by rickparrish** [comment]
I think I remember a commit in Fleck where someone enabled IPv4 and IPv6 support on the same socket, and I'm guessing that isn't supported on XP SP3.

There's other problems in FleckProxy (like my sucky thread handling) that caused me to revive some old code I was previously using for proxying websocket connections, so if you want to give that a try and let me know how it works I think it's what I'll be working on going forward:

https://github.com/rickparrish/fTelnetProxy

**Comment by HunterZ** [comment]
Thanks, I'll take a look if I have time.

For now I'm experimentally running the python-based websockify ( https://github.com/kanaka/websockify ) directly on my router to proxy for HtmlTerm, and it seems to be working fine.

Honestly it would be easiest if I could just use Synchronet's websocketservice.js WebSocket service, but HtmlTerm couldn't handshake with it for some reason. I tried a version from your site that someone linked somewhere (dated 2014, versus Synchronet's 2011 dated implementation), but it had some other error.

**Comment by rickparrish** [comment]
Yeah if you're using Synchronet, that is the ideal solution.  The problem with the 2011 version is that it pre-dates the final RFC of the WebSocket protocol, so it won't work with any current web browser.  

I'd recommend trying the latest version from CVS (http://cvs.synchro.net/cgi-bin/viewcvs.cgi/*checkout*/exec/websocketservice.js?revision=1.6&content-type=text%2Fplain) to see if that works for you.

**Comment by HunterZ** [comment]
Yeah, that's the same version I tried yesterday. I get the following error from Synchronet when I try to connect with embedded HtmlTerm:

 WebSocket ReferenceError: StringToBytes is not defined

**Comment by HunterZ** [comment]
fTelnetProxy works (after finding and installing .NET framework 2.0, which doesn't come bundled with XP) when run directly, but when I try to run it as a service, the service "terminates unexpectedly" (error 1067).

**Comment by rickparrish** [comment]
websocketservice.js reads one byte at a time, so yeah I can see it being less smooth.  I should look into whether I can read larger chunks at a time.

As for fTelnetProxy crashing, does anything get written to the log fTelnetProxy.log file?  Or anything in event viewer?  I'll have to see if I can find my old ISO and install in a VM for testing.

**Comment by HunterZ** [comment]
Does the log file get written to in service mode? I'll take a look when I get a chance.

Event viewer was useless.

**Comment by rickparrish** [comment]
Good to hear it's working now.  It's the same codebase whether you run as service or console, so I'm guessing they fixed something in the .NET framework service handling code in one of the updates you've installed.

**Comment by konchoul** [comment]
Hello and sorry for jumping in to this issue, but it also refers to WinXP as the title says.

It appears that FleckProxy no longer runs on latest upgraded Windows XP SP3 with latest .NET framework (v4.0.3).

I had a previous version up and running, but because of the AppDomain Exception searched for a newer version and downloaded it. Nothing changed, just replaced the older working .exe and .dll files and attempted to run it. It's not some exception or anything similar, the executable doesn't even run (eg fleckproxy.exe /?). I re-downloaded of course to rule out this possibility.

On a physical XP machine I just get this:
![image](https://cloud.githubusercontent.com/assets/15706606/11016403/38ca6d3a-8589-11e5-8b6a-c54233d5d6ad.png)

Trying the same on a VirtualBox XP machine just to be sure, throws this (sorry for the localized windows, I believe the type of message is recognizable though):
![image](https://cloud.githubusercontent.com/assets/15706606/11016416/8906ed28-8589-11e5-9009-476fc449c877.png)

Any ideas?

**Comment by rickparrish** [comment]
Does https://github.com/rickparrish/fTelnetProxy work?  I haven't worked on FleckProxy in awhile, and since fTelnetProxy has been working pretty good (using it on my own Windows server and half dozen Linux servers for months without issue) I probably won't be working on FleckProxy at all anymore.

fTelnetProxy will require .NET 2.0 to be installed.

**Comment by konchoul** [comment]
Thank you for your answer, unfortunately it doesn't (same 16 bit subsystem error on execution). Just to be sure, I tried with a fresh XP installation on VirtualBox. None of them even runs.

**Comment by rickparrish** [comment]
.NET apps aren't 16-bit, so I have no idea why it would be trying to execute it using the 16-bit MS-DOS subsystem.  A quick Google search revealed someone else had the same problem, but there was no resolution posted.

On a fresh XP installation it won't work, because .NET 2.0 isn't installed.  After installing that it should work.  At least in theory :)

**Comment by konchoul** [comment]
My bad, by "fresh" I meant just install the necessary and nothing extra: install XP, install .NET, try to execute it: no luck :(
So do I have any alternatives apart from changing the OS on the machine in question? It really seems that it's not anymore running on XP and my intuition says that this is because of the .dll, not the .exe.

Thank you for your time anyway for supporting this!

**Comment by rickparrish** [comment]
I just tried both scenarios, and both work for me here.  And the DLL is also built with .NET and not 16-bit, so it shouldn't be the cause of the problem either.

After some playing around I was able to duplicate the 16-bit ms-dos subsystem error by right-clicking the two files in https://github.com/rickparrish/fTelnetProxy/tree/master/bin/Release and choosing Save Target As (instead of clicking through and downloading the "raw" version).  

This downloads HTML content and saves it as the .DLL and .EXE files, and for whatever reason that causes XP to try to run the executable through the 16-bit ms-dos subsystem.

Is it possible you did that with one of the downloads, and that's why one set works and another doesn't?  The filesizes are totally different if this happened (should be 23 and 194KB, but saving the HTML yields 28 and 29KB files.  Checking the contents for text or binary would also indicate if this is what happened.  (I guess since you said they work perfectly fine on the "modern" OS this means it isn't the cause in your case, but it's the only way I could duplicate it here so I thought I'd mention it in case others run into the same problem)

Whatever the case, glad to hear it's working for you now!

**Comment by HunterZ** [comment]
Yikes. Is there a reason that you're publishing releases via the source repository instead of via Github's release feature? https://github.com/rickparrish/fTelnetProxy/releases

**Comment by konchoul** [comment]
Ouf, I really don't know what to answer on this. I guess it's github-dependent by now.
I'll just try to present the facts in case they're of any help to you, after trying everything one more time:
-Yes, for me right-clicking and "save as"  IS working on the "main/modern" OS (which shouldn't be the case taking into consideration your last post), but not on XP.
-Downloading the "raw" thing DOES always work on everything.

So eventually it seems that it all comes down to "get the raw"...

Once more, many thanks for everything! :)
